tests/integration/distribution.go

```diff
@@ -454,6 +479,398 @@ func (s *CCVTestSuite) TestSendRewardsToProvider() {
 	}
 }
 
+// TestIBCTransferMiddleware tests the logic of the IBC transfer OnRecvPacket callback
+func (s *CCVTestSuite) TestIBCTransferMiddleware() {
+
+	var (
+		data        ibctransfertypes.FungibleTokenPacketData
+		packet      channeltypes.Packet
+		getIBCDenom func(string, string) string
+	)
+
+	testCases := []struct {
+		name             string
+		setup            func(sdk.Context, *providerkeeper.Keeper, icstestingutils.TestBankKeeper)
+		rewardsAllocated bool
+		expErr           bool
+	}{
+		{
+			"invalid IBC packet",
+			func(sdk.Context, *providerkeeper.Keeper, icstestingutils.TestBankKeeper) {
+				packet = channeltypes.Packet{}
+			},
+			false,
+			true,
+		},
+		{
+			"IBC packet sender isn't a consumer chain",
+			func(ctx sdk.Context, keeper *providerkeeper.Keeper, bankKeeper icstestingutils.TestBankKeeper) {
+				// make the sender consumer chain impossible to identify
+				packet.DestinationChannel = "CorruptedChannelId"
+			},
+			false,
+			false,
+		},
+		{
+			"IBC Transfer recipient is not the consumer rewards pool address",
+			func(ctx sdk.Context, keeper *providerkeeper.Keeper, bankKeeper icstestingutils.TestBankKeeper) {
+				data.Receiver = "cosmos149lw9fktlqfed3zt8ah48r5czmsug5s7kw77u9" // random acct address
+				packet.Data = data.GetBytes()
+			},
+			false,
+			false,
+		},
+		{
+			"IBC Transfer coin denom isn't registered",
+			func(ctx sdk.Context, keeper *providerkeeper.Keeper, bankKeeper icstestingutils.TestBankKeeper) {
+				fmt.Println("middleware", getIBCDenom(packet.DestinationPort, packet.DestinationChannel))
+				keeper.SetConsumerRewardDenom(
+					s.providerCtx(),
+					getIBCDenom(packet.DestinationPort, packet.DestinationChannel),
+				)
+			},
+			false,
+			false,
+		},
+		{
+			"successful token transfer to empty pool",
+			func(ctx sdk.Context, keeper *providerkeeper.Keeper, bankKeeper icstestingutils.TestBankKeeper) {
```

Comment:
The `setup` of this test is the same as the one from the test above `"IBC Transfer coin denom isn't registered",` so it's hard for me to understand while `rewardsAllocated` is `true` in this test but `false` in the other one.